File: src/services/musicPlayer.js
import { 
  createAudioPlayer, 
  createAudioResource, 
  AudioPlayerStatus, 
  VoiceConnectionStatus,
  joinVoiceChannel,
  StreamType
} from '@discordjs/voice';
import { youtubeService } from './youtubeService.js';
import { Logger } from '../utils/logger.js';

export class MusicPlayer {
  constructor() {
    this.queues = new Map();
    this.players = new Map();
    this.connections = new Map();
    this.currentlyPlaying = new Map();
    this.isPlaying = new Map();
  }

  getQueue(guildId) {
    if (!this.queues.has(guildId)) {
      this.queues.set(guildId, []);
    }
    return this.queues.get(guildId);
  }

  getPlayer(guildId) {
    if (!this.players.has(guildId)) {
      const player = createAudioPlayer();
      this.players.set(guildId, player);
      
      player.on(AudioPlayerStatus.Idle, () => {
        Logger.info(`Player idle in guild ${guildId}, playing next song`);
        this.isPlaying.set(guildId, false);
        this.playNext(guildId);
      });

      player.on(AudioPlayerStatus.Playing, () => {
        Logger.info(`Player started playing in guild ${guildId}`);
        this.isPlaying.set(guildId, true);
      });

      player.on('error', (error) => {
        Logger.error(`Audio player error in guild ${guildId}:`, error.message);
        this.isPlaying.set(guildId, false);
        this.playNext(guildId);
      });
    }
    return this.players.get(guildId);
  }

  async joinChannel(voiceChannel) {
    try {
      const connection = joinVoiceChannel({
        channelId: voiceChannel.id,
        guildId: voiceChannel.guild.id,
        adapterCreator: voiceChannel.guild.voiceAdapterCreator,
      });

      this.connections.set(voiceChannel.guild.id, connection);

      connection.on(VoiceConnectionStatus.Disconnected, () => {
        Logger.info(`Disconnected from voice channel in guild ${voiceChannel.guild.id}`);
      });

      connection.on('error', (error) => {
        Logger.error(`Voice connection error in guild ${voiceChannel.guild.id}:`, error.message);
      });

      const player = this.getPlayer(voiceChannel.guild.id);
      connection.subscribe(player);

      return connection;
    } catch (error) {
      Logger.error('Failed to join voice channel:', error.message);
      throw error;
    }
  }

  async addToQueue(guildId, track, requestedBy) {
    const queue = this.getQueue(guildId);
    const queueItem = {
      ...track,
      requestedBy,
      addedAt: new Date()
    };
    
    queue.push(queueItem);
    Logger.info(`Added to queue: ${track.title} (Guild: ${guildId})`);
    
    // Only start playing if nothing is currently playing
    if (!this.isPlaying.get(guildId)) {
      await this.playNext(guildId);
    }
    
    return queueItem;
  }

  async addPlaylistToQueue(guildId, playlist, requestedBy) {
    const queue = this.getQueue(guildId);
    
    // Add all tracks to queue without fetching streams
    for (const track of playlist.tracks || playlist.videos) {
      const queueItem = {
        ...track,
        requestedBy,
        addedAt: new Date(),
        isPlaylist: true
      };
      queue.push(queueItem);
    }
    
    Logger.info(`Added ${playlist.tracks?.length || playlist.videos?.length} songs from playlist to queue (Guild: ${guildId})`);
    
    // Only start playing if nothing is currently playing
    if (!this.isPlaying.get(guildId)) {
      await this.playNext(guildId);
    }
    
    return playlist.tracks?.length || playlist.videos?.length;
  }

  async playNext(guildId) {
    const queue = this.getQueue(guildId);
    const player = this.getPlayer(guildId);

    if (queue.length === 0) {
      Logger.info(`Queue empty for guild ${guildId}`);
      this.isPlaying.set(guildId, false);
      this.currentlyPlaying.delete(guildId);
      return;
    }

    const track = queue.shift();
    this.currentlyPlaying.set(guildId, track);
    
    try {
      Logger.info(`Preparing to play: ${track.title} (Guild: ${guildId})`);
      
      let audioUrl = track.url;
      
      // If it's a Spotify track or search query, search for it on YouTube
      if (track.searchQuery && !track.url) {
        Logger.info(`Searching YouTube for: ${track.searchQuery}`);
        const searchResults = await youtubeService.searchVideos(track.searchQuery, 1);
        
        if (searchResults.length === 0) {
          Logger.error(`No YouTube results found for: ${track.searchQuery}`);
          return this.playNext(guildId);
        }
        
        audioUrl = searchResults[0].url;
        track.url = audioUrl;
        Logger.info(`Found YouTube URL: ${audioUrl}`);
      }
      
      if (!audioUrl) {
        Logger.error(`No URL available for track: ${track.title}`);
        return this.playNext(guildId);
      }
      
      // Get audio stream only when we're about to play
      Logger.info(`Creating audio stream for: ${audioUrl}`);
      const stream = await youtubeService.getAudioStream(audioUrl);
      
      const resource = createAudioResource(stream, {
        inputType: StreamType.Arbitrary,
        inlineVolume: true,
        metadata: {
          title: track.title,
          url: audioUrl
        }
      });
      
      player.play(resource);
      
      Logger.info(`Now playing: ${track.title} (Guild: ${guildId})`);
    } catch (error) {
      Logger.error(`Failed to play track: ${track.title}`, error.message);
      Logger.info('Attempting to play next track...');
      this.isPlaying.set(guildId, false);
      return this.playNext(guildId);
    }
  }

  skip(guildId) {
    const player = this.getPlayer(guildId);
    const queue = this.getQueue(guildId);
    
    if (this.isPlaying.get(guildId) || queue.length > 0) {
      player.stop(); // This will trigger the 'idle' event and play next song
      return true;
    }
    return false;
  }

  pause(guildId) {
    const player = this.getPlayer(guildId);
    return player.pause();
  }

  resume(guildId) {
    const player = this.getPlayer(guildId);
    return player.unpause();
  }

  stop(guildId) {
    const player = this.getPlayer(guildId);
    const queue = this.getQueue(guildId);
    
    queue.length = 0;
    this.isPlaying.set(guildId, false);
    this.currentlyPlaying.delete(guildId);
    player.stop();
    
    const connection = this.connections.get(guildId);
    if (connection) {
      connection.destroy();
      this.connections.delete(guildId);
    }
  }

  getCurrentQueue(guildId) {
    const queue = this.getQueue(guildId);
    const currentTrack = this.currentlyPlaying.get(guildId);
    
    if (currentTrack) {
      return [currentTrack, ...queue];
    }
    return queue;
  }

  getCurrentTrack(guildId) {
    return this.currentlyPlaying.get(guildId);
  }

  leave(guildId) {
    const connection = this.connections.get(guildId);
    if (connection) {
      this.stop(guildId);
      return true;
    }
    return false;
  }
}

export const musicPlayer = new MusicPlayer();

File: src/services/music_player.py
"""
Music player service for handling audio playback
"""

import discord
import asyncio
import logging
from src.services.youtube_service import YouTubeService
from src.utils.logger import get_logger

logger = get_logger(__name__)

class MusicPlayer:
    """Music player class for handling audio playback"""
    
    def __init__(self):
        self.queues = {}
        self.voice_clients = {}
        self.currently_playing = {}
        self.is_playing = {}
        self.youtube_service = YouTubeService()
    
    def get_queue(self, guild_id):
        """Get queue for a guild"""
        if guild_id not in self.queues:
            self.queues[guild_id] = []
        return self.queues[guild_id]
    
    async def join_channel(self, voice_channel):
        """Join a voice channel"""
        try:
            guild_id = voice_channel.guild.id
            
            # If already connected to a different channel, move
            if guild_id in self.voice_clients:
                await self.voice_clients[guild_id].move_to(voice_channel)
            else:
                voice_client = await voice_channel.connect()
                self.voice_clients[guild_id] = voice_client
            
            logger.info(f"Joined voice channel in guild {guild_id}")
            return self.voice_clients[guild_id]
            
        except Exception as error:
            logger.error(f"Failed to join voice channel: {error}")
            raise error
    
    async def add_to_queue(self, guild_id, track, requested_by):
        """Add a single track to the queue"""
        queue = self.get_queue(guild_id)
        
        queue_item = {
            **track,
            'requested_by': requested_by,
            'added_at': asyncio.get_event_loop().time()
        }
        
        queue.append(queue_item)
        logger.info(f"Added to queue: {track['title']} (Guild: {guild_id})")
        
        # Start playing if nothing is currently playing
        if not self.is_playing.get(guild_id, False):
            await self.play_next(guild_id)
        
        return queue_item
    
    async def add_playlist_to_queue(self, guild_id, playlist, requested_by):
        """Add a playlist to the queue"""
        queue = self.get_queue(guild_id)
        
        tracks = playlist.get('tracks', playlist.get('videos', []))
        
        for track in tracks:
            queue_item = {
                **track,
                'requested_by': requested_by,
                'added_at': asyncio.get_event_loop().time(),
                'is_playlist': True
            }
            queue.append(queue_item)
        
        logger.info(f"Added {len(tracks)} songs from playlist to queue (Guild: {guild_id})")
        
        # Start playing if nothing is currently playing
        if not self.is_playing.get(guild_id, False):
            await self.play_next(guild_id)
        
        return len(tracks)
    
    async def play_next(self, guild_id):
        """Play the next song in the queue"""
        queue = self.get_queue(guild_id)
        voice_client = self.voice_clients.get(guild_id)
        
        if not queue or not voice_client:
            logger.info(f"Queue empty or no voice client for guild {guild_id}")
            self.is_playing[guild_id] = False
            if guild_id in self.currently_playing:
                del self.currently_playing[guild_id]
            return
        
        track = queue.pop(0)
        self.currently_playing[guild_id] = track
        
        try:
            logger.info(f"Preparing to play: {track['title']} (Guild: {guild_id})")
            
            # Get audio URL
            audio_url = track.get('url')
            
            # If it's a Spotify track or search query, search for it on YouTube
            if track.get('search_query') and not audio_url:
                logger.info(f"Searching YouTube for: {track['search_query']}")
                search_results = await self.youtube_service.search_videos(track['search_query'], 1)
                
                if not search_results:
                    logger.error(f"No YouTube results found for: {track['search_query']}")
                    await self.play_next(guild_id)
                    return
                
                audio_url = search_results[0]['url']
                track['url'] = audio_url
                logger.info(f"Found YouTube URL: {audio_url}")
            
            if not audio_url:
                logger.error(f"No URL available for track: {track['title']}")
                await self.play_next(guild_id)
                return
            
            # Create audio source
            logger.info(f"Creating audio source for: {audio_url}")
            
            # Use yt-dlp to get the audio stream
            ytdl_format_options = {
                'format': 'bestaudio/best',
                'outtmpl': '%(extractor)s-%(id)s-%(title)s.%(ext)s',
                'restrictfilenames': True,
                'noplaylist': True,
                'nocheckcertificate': True,
                'ignoreerrors': False,
                'logtostderr': False,
                'quiet': True,
                'no_warnings': True,
                'default_search': 'auto',
                'source_address': '0.0.0.0'
            }
            
            ffmpeg_options = {
                'before_options': '-reconnect 1 -reconnect_streamed 1 -reconnect_delay_max 5',
                'options': '-vn'
            }
            
            audio_source = discord.FFmpegPCMAudio(
                audio_url,
                **ffmpeg_options,
                executable='ffmpeg'
            )
            
            # Play the audio
            def after_playing(error):
                if error:
                    logger.error(f"Player error: {error}")
                else:
                    logger.info(f"Finished playing: {track['title']}")
                
                # Schedule next song
                asyncio.run_coroutine_threadsafe(
                    self.play_next(guild_id),
                    asyncio.get_event_loop()
                )
            
            voice_client.play(audio_source, after=after_playing)
            self.is_playing[guild_id] = True
            
            logger.info(f"Now playing: {track['title']} (Guild: {guild_id})")
            
        except Exception as error:
            logger.error(f"Failed to play track: {track['title']} - {error}")
            logger.info("Attempting to play next track...")
            self.is_playing[guild_id] = False
            await self.play_next(guild_id)
    
    def skip(self, guild_id):
        """Skip the current song"""
        voice_client = self.voice_clients.get(guild_id)
        
        if voice_client and voice_client.is_playing():
            voice_client.stop()
            return True
        return False
    
    def pause(self, guild_id):
        """Pause the current song"""
        voice_client = self.voice_clients.get(guild_id)
        
        if voice_client and voice_client.is_playing():
            voice_client.pause()
            return True
        return False
    
    def resume(self, guild_id):
        """Resume the current song"""
        voice_client = self.voice_clients.get(guild_id)
        
        if voice_client and voice_client.is_paused():
            voice_client.resume()
            return True
        return False
    
    def stop(self, guild_id):
        """Stop music and clear queue"""
        queue = self.get_queue(guild_id)
        voice_client = self.voice_clients.get(guild_id)
        
        queue.clear()
        self.is_playing[guild_id] = False
        
        if guild_id in self.currently_playing:
            del self.currently_playing[guild_id]
        
        if voice_client:
            voice_client.stop()
    
    def get_current_queue(self, guild_id):
        """Get the current queue including currently playing song"""
        queue = self.get_queue(guild_id)
        current_track = self.currently_playing.get(guild_id)
        
        if current_track:
            return [current_track] + queue
        return queue
    
    def get_current_track(self, guild_id):
        """Get the currently playing track"""
        return self.currently_playing.get(guild_id)
    
    async def leave(self, guild_id):
        """Leave the voice channel"""
        voice_client = self.voice_clients.get(guild_id)
        
        if voice_client:
            self.stop(guild_id)
            await voice_client.disconnect()
            del self.voice_clients[guild_id]
            return True
        return False